pkg/uefi/meregion.go

```diff
@@ -0,0 +1,238 @@
+	}
+	fp := &MEFPT{PartitionMapStart: o + 28}
+	r := bytes.NewReader(buf[o:])
+	if err := binary.Read(r, binary.LittleEndian, &fp.PartitionCount); err != nil {
+		return nil, err
+	}
+	l := fp.PartitionMapStart + MEPartitionTableEntryLength*int(fp.PartitionCount)
+	if len(buf) < l {
+		return nil, fmt.Errorf("ME section (%#x) too small for %d entries in ME Flash Partition Table (%#x)", len(buf), fp.PartitionCount, l)
+	}
+
+	fp.buf = make([]byte, l)
+	copy(fp.buf, buf[:l])
+	if err := fp.parsePartitions(); err != nil {
+		return nil, err
+	}
+	return fp, nil
+}
+
+func (fp *MEFPT) parsePartitions() error {
+	r := bytes.NewReader(fp.buf[fp.PartitionMapStart:])
+	for i := 0; i < int(fp.PartitionCount); i++ {
+		var entry MePartitionEntry
+		if err := binary.Read(r, binary.LittleEndian, &entry); err != nil {
+			return err
+		}
+		fp.Entries = append(fp.Entries, entry)
+	}
+	return nil
+}
+
+// MERegion implements Region for a raw chunk of bytes in the firmware image.
+type MERegion struct {
+	FPT *MEFPT
+	// holds the raw data
+	buf []byte
+	// Metadata for extraction and recovery
+	ExtractPath string
+	// This is a pointer to the FlashRegion struct laid out in the ifd.
+	FRegion *FlashRegion
+	// Region Type as per the IFD
+	RegionType FlashRegionType
+	// Computed free space after parsing the partition table
+	FreeSpaceOffset uint64
+}
+
+// SetFlashRegion sets the flash region.
+func (rr *MERegion) SetFlashRegion(fr *FlashRegion) {
+	rr.FRegion = fr
+}
+
+// FlashRegion gets the flash region.
+func (rr *MERegion) FlashRegion() (fr *FlashRegion) {
+	return rr.FRegion
+}
+
+// NewMERegion creates a new region.
+func NewMERegion(buf []byte, r *FlashRegion, rt FlashRegionType) (Region, error) {
+	rr := &MERegion{FRegion: r, RegionType: rt}
+	rr.buf = make([]byte, len(buf))
+	copy(rr.buf, buf)
+	fp, err := NewMEFPT(buf)
+	if err != nil {
+		log.Printf("error parsing ME Flash Partition Table: %v", err)
+		return rr, nil
+	}
+	rr.FPT = fp
+	// Compute FreeSpaceOffset
+	for _, p := range fp.Entries {
+		endOffset := uint64(p.Offset) + uint64(p.Length)
+		if endOffset > rr.FreeSpaceOffset {
+			rr.FreeSpaceOffset = endOffset
+		}
+	}
+
+	return rr, nil
+}
+
+// Type returns the flash region type.
+func (rr *MERegion) Type() FlashRegionType {
+	return RegionTypeME
+}
+
+// Buf returns the buffer.
+// Used mostly for things interacting with the Firmware interface.
+func (rr *MERegion) Buf() []byte {
+	return rr.buf
+}
+
+// SetBuf sets the buffer.
+// Used mostly for things interacting with the Firmware interface.
+func (rr *MERegion) SetBuf(buf []byte) {
+	rr.buf = buf
+}
+
+// Apply calls the visitor on the MERegion.
+func (rr *MERegion) Apply(v Visitor) error {
+	return v.Visit(rr)
+}
+
+// ApplyChildren calls the visitor on each child node of MERegion.
+func (rr *MERegion) ApplyChildren(v Visitor) error {
+	if rr.FPT == nil {
+		return nil
+	}
+	if err := rr.FPT.Apply(v); err != nil {
```

Comment:
This can be simplified to "return rr.FPT.Apply(v)". You don't need the if-statement.

Reply:
Right, thanks